azure/services/availabilitysets/client.go

```diff
@@ -19,83 +19,78 @@ package availabilitysets
-func NewClient(auth azure.Authorizer) *AzureClient {
-	return &AzureClient{
-		availabilitySets: newAvailabilitySetsClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer()),
+// NewClient creates a new availability sets client from an authorizer.
+func NewClient(auth azure.Authorizer) (*AzureClient, error) {
+	opts, err := azure.ARMClientOptions(auth.CloudEnvironment())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create availabilitysets client options")
 	}
-}
-
-// newAvailabilitySetsClient creates a new AvailabilitySets Client from subscription ID.
-func newAvailabilitySetsClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) compute.AvailabilitySetsClient {
-	asClient := compute.NewAvailabilitySetsClientWithBaseURI(baseURI, subscriptionID)
-	azure.SetAutoRestClientDefaults(&asClient.Client, authorizer)
-	return asClient
+	factory, err := armcompute.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create armcompute client factory")
+	}
+	return &AzureClient{factory.NewAvailabilitySetsClient()}, nil
 }
 
 // Get gets an availability set.
 func (ac *AzureClient) Get(ctx context.Context, spec azure.ResourceSpecGetter) (result interface{}, err error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "availabilitysets.AzureClient.Get")
 	defer done()
 
-	return ac.availabilitySets.Get(ctx, spec.ResourceGroupName(), spec.ResourceName())
+	resp, err := ac.availabilitySets.Get(ctx, spec.ResourceGroupName(), spec.ResourceName(), nil)
+	if err != nil {
+		return nil, err
+	}
+	return resp.AvailabilitySet, nil
 }
 
-// CreateOrUpdateAsync creates or updates a availability set asynchronously.
-// It sends a PUT request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// CreateOrUpdateAsync creates or updates an availability set asynchronously.
+// It sends a PUT request to Azure and if accepted without error, the func will return a Poller which can be used to track the ongoing
 // progress of the operation.
-func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{}) (result interface{}, future azureautorest.FutureAPI, err error) {
+func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, _resumeToken string, parameters interface{}) (result interface{}, poller *runtime.Poller[armcompute.AvailabilitySetsClientCreateOrUpdateResponse], err error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "availabilitySets.AzureClient.CreateOrUpdateAsync")
 	defer done()
 
-	availabilitySet, ok := parameters.(compute.AvailabilitySet)
+	availabilitySet, ok := parameters.(armcompute.AvailabilitySet)
 	if !ok {
-		return nil, nil, errors.Errorf("%T is not a compute.AvailabilitySet", parameters)
+		return nil, nil, errors.Errorf("%T is not an armcompute.AvailabilitySet", parameters)
```

Comment:
Adding it as a placeholder, we probably also want to check of `parameters` is not `nil` before returning `nil, nil, errors.Errorf("%T is not an armcompute.AvailabilitySet", parameters)`?

Reply:
You mean like this? https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/b0829812e2ade8850c7023442cd17fa6dcabed51/azure/services/availabilitysets/client.go#L67

Reply:
Ou yep. It is already in there.